// symbol-openapi-typescript-fetch-client/dist/apis/FinalizationRoutesApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FinalizationRoutesApi = void 0;
const runtime = require("../runtime");
const models_1 = require("../models");
/**
 *
 */
class FinalizationRoutesApi extends runtime.BaseAPI {
    /**
     * Gets finalization proof for the greatest height associated with the given epoch.
     * Get finalization proof
     */
    getFinalizationProofAtEpochRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.epoch === null || requestParameters.epoch === undefined) {
                throw new runtime.RequiredError('epoch', 'Required parameter requestParameters.epoch was null or undefined when calling getFinalizationProofAtEpoch.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/finalization/proof/epoch/{epoch}`.replace(`{${"epoch"}}`, encodeURIComponent(String(requestParameters.epoch))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.FinalizationProofDTOFromJSON(jsonValue));
        });
    }
    /**
     * Gets finalization proof for the greatest height associated with the given epoch.
     * Get finalization proof
     */
    getFinalizationProofAtEpoch(epoch) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getFinalizationProofAtEpochRaw({ epoch: epoch });
            return yield response.value();
        });
    }
    /**
     * Gets finalization proof at the given height.
     * Get finalization proof
     */
    getFinalizationProofAtHeightRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.height === null || requestParameters.height === undefined) {
                throw new runtime.RequiredError('height', 'Required parameter requestParameters.height was null or undefined when calling getFinalizationProofAtHeight.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/finalization/proof/height/{height}`.replace(`{${"height"}}`, encodeURIComponent(String(requestParameters.height))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.FinalizationProofDTOFromJSON(jsonValue));
        });
    }
    /**
     * Gets finalization proof at the given height.
     * Get finalization proof
     */
    getFinalizationProofAtHeight(height) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getFinalizationProofAtHeightRaw({ height: height });
            return yield response.value();
        });
    }
}
exports.FinalizationRoutesApi = FinalizationRoutesApi;
