// symbol-sdk/dist/src/model/restriction/AccountRestrictions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountRestrictions = void 0;
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const catbuffer_typescript_1 = require("catbuffer-typescript");
const internal_compatibility_1 = require("rxjs/internal-compatibility");
const account_1 = require("../account");
const mosaic_1 = require("../mosaic");
/**
 * Account restrictions structure describes restriction information for an account.
 */
class AccountRestrictions {
    /**
     * Constructor
     * @param version version
     * @param recordId the data base id.
     * @param address the target address
     * @param restrictions the restrictions
     */
    constructor(version, recordId, address, restrictions) {
        this.version = version;
        this.recordId = recordId;
        this.address = address;
        this.restrictions = restrictions;
    }
    /**
     * Generate buffer
     * @return {Uint8Array}
     */
    serialize() {
        const address = this.address.toBuilder();
        const restrictions = this.restrictions.map((r) => {
            const addressRestrictions = new catbuffer_typescript_1.AccountRestrictionAddressValueBuilder(r.values.filter((v) => v instanceof account_1.Address).map((a) => a.toBuilder()));
            const mosaicIdRestrictions = new catbuffer_typescript_1.AccountRestrictionMosaicValueBuilder(r.values.filter((v) => v instanceof mosaic_1.MosaicId).map((a) => a.toBuilder()));
            const transactionTypeRestrictions = new catbuffer_typescript_1.AccountRestrictionTransactionTypeValueBuilder(r.values.filter((v) => internal_compatibility_1.isNumeric(v)).map((a) => a));
            return new catbuffer_typescript_1.AccountRestrictionsInfoBuilder(catbuffer_typescript_1.GeneratorUtils.toFlags(catbuffer_typescript_1.AccountRestrictionFlagsDto, r.restrictionFlags), addressRestrictions, mosaicIdRestrictions, transactionTypeRestrictions);
        });
        return new catbuffer_typescript_1.AccountRestrictionsBuilder(this.version, address, restrictions).serialize();
    }
}
exports.AccountRestrictions = AccountRestrictions;
