// symbol-sdk/dist/src/model/restriction/MosaicAddressRestriction.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicAddressRestriction = void 0;
const catbuffer_typescript_1 = require("catbuffer-typescript");
class MosaicAddressRestriction {
    /**
     * Constructor
     * @param version
     * @param compositeHash
     * @param entryType
     * @param mosaicId
     * @param targetAddress
     * @param restrictions
     */
    constructor(
    /**
     * Version
     */
    version, 
    /**
     * composite hash
     */
    compositeHash, 
    /**
     * Mosaic restriction entry type.
     */
    entryType, 
    /**
     * Mosaic identifier.
     */
    mosaicId, 
    /**
     * Target address
     */
    targetAddress, 
    /**
     * Mosaic restriction items
     */
    restrictions) {
        this.version = version;
        this.compositeHash = compositeHash;
        this.entryType = entryType;
        this.mosaicId = mosaicId;
        this.targetAddress = targetAddress;
        this.restrictions = restrictions;
    }
    /**
     * Returns the restriction for a given key.
     *
     * @param key the key.
     */
    getRestriction(key) {
        return this.restrictions.find((item) => item.key.equals(key));
    }
    /**
     * Generate buffer
     * @return {Uint8Array}
     */
    serialize() {
        const mosaicId = this.mosaicId.toBuilder();
        const address = this.targetAddress.toBuilder();
        const keyPairs = new catbuffer_typescript_1.AddressKeyValueSetBuilder(this.restrictions
            .sort((a, b) => a.key.compare(b.key))
            .map((item) => {
            const key = new catbuffer_typescript_1.MosaicRestrictionKeyDto(item.key.toDTO());
            const value = item.restrictionValue.toDTO();
            return new catbuffer_typescript_1.AddressKeyValueBuilder(key, value);
        }));
        const addressRestrictionBuilder = new catbuffer_typescript_1.MosaicAddressRestrictionEntryBuilder(mosaicId, address, keyPairs);
        return new catbuffer_typescript_1.MosaicRestrictionEntryBuilder(this.version, catbuffer_typescript_1.MosaicRestrictionEntryTypeDto.ADDRESS, addressRestrictionBuilder, undefined).serialize();
    }
}
exports.MosaicAddressRestriction = MosaicAddressRestriction;
