// symbol-sdk/dist/src/infrastructure/RestrictionMosaicHttp.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestrictionMosaicHttp = void 0;
const symbol_openapi_typescript_fetch_client_1 = require("symbol-openapi-typescript-fetch-client");
const utils_1 = require("../core/utils");
const model_1 = require("../model");
const account_1 = require("../model/account");
const mosaic_1 = require("../model/mosaic");
const restriction_1 = require("../model/restriction");
const Http_1 = require("./Http");
const paginationStreamer_1 = require("./paginationStreamer");
/**
 * RestrictionMosaic http repository.
 *
 * @since 1.0
 */
class RestrictionMosaicHttp extends Http_1.Http {
    /**
     * Constructor
     * @param url Base catapult-rest url
     * @param fetchApi fetch function to be used when performing rest requests.
     */
    constructor(url, fetchApi) {
        super(url, fetchApi);
        this.restrictionMosaicRoutesApi = new symbol_openapi_typescript_fetch_client_1.RestrictionMosaicRoutesApi(this.config());
    }
    /**
     * Returns a mosaic restrictions page based on the criteria.
     *
     * @param criteria the criteria
     * @return a page of {@link MosaicRestriction}
     */
    search(criteria) {
        var _a, _b, _c;
        return this.call(this.restrictionMosaicRoutesApi.searchMosaicRestrictions((_a = criteria.mosaicId) === null || _a === void 0 ? void 0 : _a.toHex(), (_b = criteria.entryType) === null || _b === void 0 ? void 0 : _b.valueOf(), (_c = criteria.targetAddress) === null || _c === void 0 ? void 0 : _c.plain(), criteria.pageSize, criteria.pageNumber, criteria.offset, utils_1.DtoMapping.mapEnum(criteria.order)), (body) => super.toPage(body.pagination, body.data, (r) => RestrictionMosaicHttp.toMosaicRestriction(r)));
    }
    streamer() {
        return new paginationStreamer_1.RestrictionMosaicPaginationStreamer(this);
    }
    /**
     * This method maps a mosaic restriction dto from rest to the SDK's model object.
     *
     * @internal
     * @param {MosaicAddressRestrictionDTO | MosaicGlobalRestrictionDTO} dto the restriction object from rest.
     * @returns {MosaicRestriction} a restriction model
     */
    static toMosaicRestriction(dto) {
        if (dto.mosaicRestrictionEntry.targetAddress) {
            const addressRestrictionDto = dto;
            return new restriction_1.MosaicAddressRestriction(dto.mosaicRestrictionEntry.version || 1, dto.mosaicRestrictionEntry.compositeHash, dto.mosaicRestrictionEntry.entryType.valueOf(), new mosaic_1.MosaicId(dto.mosaicRestrictionEntry.mosaicId), account_1.Address.createFromEncoded(addressRestrictionDto.mosaicRestrictionEntry.targetAddress), addressRestrictionDto.mosaicRestrictionEntry.restrictions.map(RestrictionMosaicHttp.toMosaicAddressRestrictionItem));
        }
        const globalRestrictionDto = dto;
        return new restriction_1.MosaicGlobalRestriction(dto.mosaicRestrictionEntry.version || 1, dto.mosaicRestrictionEntry.compositeHash, dto.mosaicRestrictionEntry.entryType.valueOf(), new mosaic_1.MosaicId(dto.mosaicRestrictionEntry.mosaicId), globalRestrictionDto.mosaicRestrictionEntry.restrictions.map((i) => RestrictionMosaicHttp.toMosaicGlobalRestrictionItem(i)));
    }
    static toMosaicGlobalRestrictionItem(restriction) {
        return new restriction_1.MosaicGlobalRestrictionItem(model_1.UInt64.fromNumericString(restriction.key), new mosaic_1.MosaicId(restriction.restriction.referenceMosaicId), model_1.UInt64.fromNumericString(restriction.restriction.restrictionValue), restriction.restriction.restrictionType.valueOf());
    }
    static toMosaicAddressRestrictionItem(restriction) {
        return new restriction_1.MosaicAddressRestrictionItem(model_1.UInt64.fromNumericString(restriction.key), model_1.UInt64.fromNumericString(restriction.value));
    }
    getMosaicRestrictions(compositeHash) {
        return this.call(this.restrictionMosaicRoutesApi.getMosaicRestrictions(compositeHash), RestrictionMosaicHttp.toMosaicRestriction);
    }
    getMosaicRestrictionsMerkle(compositeHash) {
        return this.call(this.restrictionMosaicRoutesApi.getMosaicRestrictionsMerkle(compositeHash), utils_1.DtoMapping.toMerkleStateInfo);
    }
}
exports.RestrictionMosaicHttp = RestrictionMosaicHttp;
