// symbol-openapi-typescript-fetch-client/src/apis/RestrictionMosaicRoutesApi.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import {
    MerkleStateInfoDTO,
    MerkleStateInfoDTOFromJSON,
    MerkleStateInfoDTOToJSON,
    ModelError,
    ModelErrorFromJSON,
    ModelErrorToJSON,
    MosaicRestrictionDTO,
    MosaicRestrictionDTOFromJSON,
    MosaicRestrictionDTOToJSON,
    MosaicRestrictionEntryTypeEnum,
    MosaicRestrictionEntryTypeEnumFromJSON,
    MosaicRestrictionEntryTypeEnumToJSON,
    MosaicRestrictionsPage,
    MosaicRestrictionsPageFromJSON,
    MosaicRestrictionsPageToJSON,
    Order,
    OrderFromJSON,
    OrderToJSON,
} from '../models';

export interface GetMosaicRestrictionsRequest {
    compositeHash: string;
}

export interface GetMosaicRestrictionsMerkleRequest {
    compositeHash: string;
}

export interface SearchMosaicRestrictionsRequest {
    mosaicId?: string;
    entryType?: MosaicRestrictionEntryTypeEnum;
    targetAddress?: string;
    pageSize?: number;
    pageNumber?: number;
    offset?: string;
    order?: Order;
}

/**
 * 
 */
export class RestrictionMosaicRoutesApi extends runtime.BaseAPI {

    /**
     * Returns the mosaic restrictions for a composite hash.
     * Get the mosaic restrictions
     */
    async getMosaicRestrictionsRaw(requestParameters: GetMosaicRestrictionsRequest): Promise<runtime.ApiResponse<MosaicRestrictionDTO>> {
        if (requestParameters.compositeHash === null || requestParameters.compositeHash === undefined) {
            throw new runtime.RequiredError('compositeHash','Required parameter requestParameters.compositeHash was null or undefined when calling getMosaicRestrictions.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/restrictions/mosaic/{compositeHash}`.replace(`{${"compositeHash"}}`, encodeURIComponent(String(requestParameters.compositeHash))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => MosaicRestrictionDTOFromJSON(jsonValue));
    }

    /**
     * Returns the mosaic restrictions for a composite hash.
     * Get the mosaic restrictions
     */
    async getMosaicRestrictions(compositeHash: string): Promise<MosaicRestrictionDTO> {
        const response = await this.getMosaicRestrictionsRaw({ compositeHash: compositeHash });
        return await response.value();
    }

    /**
     * Returns the mosaic restrictions merkle for a given composite hash.
     * Get the mosaic restrictions merkle
     */
    async getMosaicRestrictionsMerkleRaw(requestParameters: GetMosaicRestrictionsMerkleRequest): Promise<runtime.ApiResponse<MerkleStateInfoDTO>> {
        if (requestParameters.compositeHash === null || requestParameters.compositeHash === undefined) {
            throw new runtime.RequiredError('compositeHash','Required parameter requestParameters.compositeHash was null or undefined when calling getMosaicRestrictionsMerkle.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/restrictions/mosaic/{compositeHash}/merkle`.replace(`{${"compositeHash"}}`, encodeURIComponent(String(requestParameters.compositeHash))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => MerkleStateInfoDTOFromJSON(jsonValue));
    }

    /**
     * Returns the mosaic restrictions merkle for a given composite hash.
     * Get the mosaic restrictions merkle
     */
    async getMosaicRestrictionsMerkle(compositeHash: string): Promise<MerkleStateInfoDTO> {
        const response = await this.getMosaicRestrictionsMerkleRaw({ compositeHash: compositeHash });
        return await response.value();
    }

    /**
     * Returns an array of mosaic restrictions.
     * Search mosaic restrictions
     */
    async searchMosaicRestrictionsRaw(requestParameters: SearchMosaicRestrictionsRequest): Promise<runtime.ApiResponse<MosaicRestrictionsPage>> {
        const queryParameters: runtime.HTTPQuery = {};

        if (requestParameters.mosaicId !== undefined) {
            queryParameters['mosaicId'] = requestParameters.mosaicId;
        }

        if (requestParameters.entryType !== undefined) {
            queryParameters['entryType'] = requestParameters.entryType;
        }

        if (requestParameters.targetAddress !== undefined) {
            queryParameters['targetAddress'] = requestParameters.targetAddress;
        }

        if (requestParameters.pageSize !== undefined) {
            queryParameters['pageSize'] = requestParameters.pageSize;
        }

        if (requestParameters.pageNumber !== undefined) {
            queryParameters['pageNumber'] = requestParameters.pageNumber;
        }

        if (requestParameters.offset !== undefined) {
            queryParameters['offset'] = requestParameters.offset;
        }

        if (requestParameters.order !== undefined) {
            queryParameters['order'] = requestParameters.order;
        }

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/restrictions/mosaic`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => MosaicRestrictionsPageFromJSON(jsonValue));
    }

    /**
     * Returns an array of mosaic restrictions.
     * Search mosaic restrictions
     */
    async searchMosaicRestrictions(mosaicId?: string, entryType?: MosaicRestrictionEntryTypeEnum, targetAddress?: string, pageSize?: number, pageNumber?: number, offset?: string, order?: Order): Promise<MosaicRestrictionsPage> {
        const response = await this.searchMosaicRestrictionsRaw({ mosaicId: mosaicId, entryType: entryType, targetAddress: targetAddress, pageSize: pageSize, pageNumber: pageNumber, offset: offset, order: order });
        return await response.value();
    }

}
