// symbol-sdk/dist/src/model/mosaic/Currency.js
"use strict";
/*
 * Copyright 2020 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Currency = void 0;
const NamespaceId_1 = require("../namespace/NamespaceId");
const UInt64_1 = require("../UInt64");
const Mosaic_1 = require("./Mosaic");
/**
 * An object that knows how to create Mosaics based on the Mosaic Info and Namespace configuration.
 */
class Currency {
    constructor({ unresolvedMosaicId, mosaicId, namespaceId, divisibility, transferable, supplyMutable, restrictable, }) {
        // If unresolvedMosaicId is not provided explicitly, mosaic id wins over namespace id for performace reasons.
        const finalMosaicId = unresolvedMosaicId || mosaicId || namespaceId;
        if (!finalMosaicId) {
            throw new Error('At least one Mosaic Id or Namespace Id must be provided');
        }
        this.unresolvedMosaicId = finalMosaicId;
        this.mosaicId = mosaicId;
        this.namespaceId = namespaceId;
        this.divisibility = divisibility;
        this.transferable = transferable;
        this.supplyMutable = supplyMutable;
        this.restrictable = restrictable;
    }
    /**
     * Creates a Mosaic from this relative amount.
     *
     * @param amount
     * @returns {Mosaic}
     */
    createRelative(amount) {
        if (typeof amount === 'number') {
            return new Mosaic_1.Mosaic(this.unresolvedMosaicId, UInt64_1.UInt64.fromUint(amount * Math.pow(10, this.divisibility)));
        }
        return new Mosaic_1.Mosaic(this.unresolvedMosaicId, UInt64_1.UInt64.fromUint(amount.compact() * Math.pow(10, this.divisibility)));
    }
    /**
     * Creates a Mosaic from this relative amount.
     *
     * @param amount
     * @returns {Mosaic}
     */
    createAbsolute(amount) {
        if (typeof amount === 'number') {
            return new Mosaic_1.Mosaic(this.unresolvedMosaicId, UInt64_1.UInt64.fromUint(amount));
        }
        return new Mosaic_1.Mosaic(this.unresolvedMosaicId, amount);
    }
}
exports.Currency = Currency;
/**
 * Currency for public / Public_test network.
 *
 * This represents the per-network currency mosaic. This mosaicId is aliased with namespace name `symbol.xym`.
 *
 * This simplifies offline operations but general applications should load the currency from the repository factory and network currency service.
 *
 * If you are creating a private network and you need offline access, you can create a Currency in memory.
 *
 */
Currency.PUBLIC = new Currency({
    namespaceId: new NamespaceId_1.NamespaceId('symbol.xym'),
    divisibility: 6,
    transferable: true,
    supplyMutable: false,
    restrictable: false,
});
