// symbol-sdk/dist/src/model/mosaic/MosaicInfo.js
"use strict";
/*
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicInfo = void 0;
const catbuffer_typescript_1 = require("catbuffer-typescript");
/**
 * The mosaic info structure describes a mosaic.
 */
class MosaicInfo {
    /**
     * @param version
     * @param recordId
     * @param id
     * @param supply
     * @param startHeight
     * @param ownerAddress
     * @param revision
     * @param flags
     * @param divisibility
     * @param duration
     */
    constructor(
    /**
     * Version
     */
    version, 
    /**
     * The database record id.
     */
    recordId, 
    /**
     * The mosaic id.
     */
    id, 
    /**
     * The mosaic supply.
     */
    supply, 
    /**
     * The block height were mosaic was created.
     */
    startHeight, 
    /**
     * The mosaic owner address.
     */
    ownerAddress, 
    /**
     * The mosaic revision
     */
    revision, 
    /**
     * The mosaic flags.
     */
    flags, 
    /**
     * Mosaic divisibility
     */
    divisibility, 
    /**
     * Mosaic duration
     */
    duration) {
        this.version = version;
        this.recordId = recordId;
        this.id = id;
        this.supply = supply;
        this.startHeight = startHeight;
        this.ownerAddress = ownerAddress;
        this.revision = revision;
        this.flags = flags;
        this.divisibility = divisibility;
        this.duration = duration;
    }
    /**
     * Is mosaic supply mutable
     * @returns {boolean}
     */
    isSupplyMutable() {
        return this.flags.supplyMutable;
    }
    /**
     * Is mosaic transferable
     * @returns {boolean}
     */
    isTransferable() {
        return this.flags.transferable;
    }
    /**
     * Is mosaic restrictable
     * @returns {boolean}
     */
    isRestrictable() {
        return this.flags.restrictable;
    }
    /**
     * Generate buffer
     * @return {Uint8Array}
     */
    serialize() {
        const mosaicId = this.id.toBuilder();
        const supply = new catbuffer_typescript_1.AmountDto(this.supply.toDTO());
        const startHeight = new catbuffer_typescript_1.HeightDto(this.startHeight.toDTO());
        const ownerAddress = this.ownerAddress.toBuilder();
        const revision = this.revision;
        const duration = new catbuffer_typescript_1.BlockDurationDto(this.duration.toDTO());
        const flags = catbuffer_typescript_1.GeneratorUtils.toFlags(catbuffer_typescript_1.MosaicFlagsDto, this.flags.getValue());
        const properties = new catbuffer_typescript_1.MosaicPropertiesBuilder(flags, this.divisibility, duration);
        const definition = new catbuffer_typescript_1.MosaicDefinitionBuilder(startHeight, ownerAddress, revision, properties);
        return new catbuffer_typescript_1.MosaicEntryBuilder(this.version, mosaicId, supply, definition).serialize();
    }
}
exports.MosaicInfo = MosaicInfo;
