// symbol-openapi-typescript-fetch-client/src/models/BmTreeSignature.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    ParentPublicKeySignaturePair,
    ParentPublicKeySignaturePairFromJSON,
    ParentPublicKeySignaturePairFromJSONTyped,
    ParentPublicKeySignaturePairToJSON,
} from './';

/**
 * 
 * @export
 * @interface BmTreeSignature
 */
export interface BmTreeSignature {
    /**
     * 
     * @type {ParentPublicKeySignaturePair}
     * @memberof BmTreeSignature
     */
    root: ParentPublicKeySignaturePair;
    /**
     * 
     * @type {ParentPublicKeySignaturePair}
     * @memberof BmTreeSignature
     */
    bottom: ParentPublicKeySignaturePair;
}

export function BmTreeSignatureFromJSON(json: any): BmTreeSignature {
    return BmTreeSignatureFromJSONTyped(json, false);
}

export function BmTreeSignatureFromJSONTyped(json: any, ignoreDiscriminator: boolean): BmTreeSignature {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'root': json['root'],
        'bottom': json['bottom'],
    };
}

export function BmTreeSignatureToJSON(value?: BmTreeSignature | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'root': value.root,
        'bottom': value.bottom,
    };
}
