// symbol-openapi-typescript-fetch-client/src/models/LockStatus.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * Possible status of lock states:
 * * 0 - UNUSED.
 * * 1 - USED.
 * @export
 * @enum {string}
 */
export enum LockStatus {
    NUMBER_0 = 0,
    NUMBER_1 = 1
}

export function LockStatusFromJSON(json: any): LockStatus {
    return LockStatusFromJSONTyped(json, false);
}

export function LockStatusFromJSONTyped(json: any, ignoreDiscriminator: boolean): LockStatus {
    return json as LockStatus;
}

export function LockStatusToJSON(value?: LockStatus | null): any {
    return value as any;
}
