// symbol-openapi-typescript-fetch-client/src/models/MerkleTreeBranchDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MerkleTreeBranchLinkDTO,
    MerkleTreeBranchLinkDTOFromJSON,
    MerkleTreeBranchLinkDTOFromJSONTyped,
    MerkleTreeBranchLinkDTOToJSON,
    MerkleTreeNodeTypeEnum,
    MerkleTreeNodeTypeEnumFromJSON,
    MerkleTreeNodeTypeEnumFromJSONTyped,
    MerkleTreeNodeTypeEnumToJSON,
} from './';

/**
 * Merkle tree branch node.
 * @export
 * @interface MerkleTreeBranchDTO
 */
export interface MerkleTreeBranchDTO {
    /**
     * 
     * @type {MerkleTreeNodeTypeEnum}
     * @memberof MerkleTreeBranchDTO
     */
    type: MerkleTreeNodeTypeEnum;
    /**
     * Branch link path.
     * @type {string}
     * @memberof MerkleTreeBranchDTO
     */
    path: string;
    /**
     * Encoded branch link path.
     * @type {string}
     * @memberof MerkleTreeBranchDTO
     */
    encodedPath: string;
    /**
     * Nibble count.
     * @type {number}
     * @memberof MerkleTreeBranchDTO
     */
    nibbleCount: number;
    /**
     * Branch link bitmask.
     * @type {string}
     * @memberof MerkleTreeBranchDTO
     */
    linkMask: string;
    /**
     * Branch links (max 16).
     * @type {Array<MerkleTreeBranchLinkDTO>}
     * @memberof MerkleTreeBranchDTO
     */
    links: Array<MerkleTreeBranchLinkDTO>;
    /**
     * 
     * @type {string}
     * @memberof MerkleTreeBranchDTO
     */
    branchHash: string;
}

export function MerkleTreeBranchDTOFromJSON(json: any): MerkleTreeBranchDTO {
    return MerkleTreeBranchDTOFromJSONTyped(json, false);
}

export function MerkleTreeBranchDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MerkleTreeBranchDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'type': json['type'],
        'path': json['path'],
        'encodedPath': json['encodedPath'],
        'nibbleCount': json['nibbleCount'],
        'linkMask': json['linkMask'],
        'links': ((json['links'] as Array<any>)),
        'branchHash': json['branchHash'],
    };
}

export function MerkleTreeBranchDTOToJSON(value?: MerkleTreeBranchDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'type': value.type,
        'path': value.path,
        'encodedPath': value.encodedPath,
        'nibbleCount': value.nibbleCount,
        'linkMask': value.linkMask,
        'links': ((value.links as Array<any>)),
        'branchHash': value.branchHash,
    };
}
