// symbol-openapi-typescript-fetch-client/src/models/MerkleTreeLeafDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MerkleTreeNodeTypeEnum,
    MerkleTreeNodeTypeEnumFromJSON,
    MerkleTreeNodeTypeEnumFromJSONTyped,
    MerkleTreeNodeTypeEnumToJSON,
} from './';

/**
 * Merkle tree leaf node.
 * @export
 * @interface MerkleTreeLeafDTO
 */
export interface MerkleTreeLeafDTO {
    /**
     * 
     * @type {MerkleTreeNodeTypeEnum}
     * @memberof MerkleTreeLeafDTO
     */
    type: MerkleTreeNodeTypeEnum;
    /**
     * Leaf path.
     * @type {string}
     * @memberof MerkleTreeLeafDTO
     */
    path: string;
    /**
     * Encoded leaf path.
     * @type {string}
     * @memberof MerkleTreeLeafDTO
     */
    encodedPath: string;
    /**
     * Nibble count.
     * @type {number}
     * @memberof MerkleTreeLeafDTO
     */
    nibbleCount: number;
    /**
     * Leaf value (sha256 hash).
     * @type {string}
     * @memberof MerkleTreeLeafDTO
     */
    value: string;
    /**
     * 
     * @type {string}
     * @memberof MerkleTreeLeafDTO
     */
    leafHash: string;
}

export function MerkleTreeLeafDTOFromJSON(json: any): MerkleTreeLeafDTO {
    return MerkleTreeLeafDTOFromJSONTyped(json, false);
}

export function MerkleTreeLeafDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MerkleTreeLeafDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'type': json['type'],
        'path': json['path'],
        'encodedPath': json['encodedPath'],
        'nibbleCount': json['nibbleCount'],
        'value': json['value'],
        'leafHash': json['leafHash'],
    };
}

export function MerkleTreeLeafDTOToJSON(value?: MerkleTreeLeafDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'type': value.type,
        'path': value.path,
        'encodedPath': value.encodedPath,
        'nibbleCount': value.nibbleCount,
        'value': value.value,
        'leafHash': value.leafHash,
    };
}
