// symbol-openapi-typescript-fetch-client/src/models/StageEnum.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.3.0
 *
 * The version of the OpenAPI document: 0.11.1
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * Type of stage:
 * * 0 - Prevote.
 * * 1 - Precommit.
 * * 2 - Count.
 * @export
 * @enum {string}
 */
export enum StageEnum {
    NUMBER_0 = 0,
    NUMBER_1 = 1,
    NUMBER_2 = 2
}

export function StageEnumFromJSON(json: any): StageEnum {
    return StageEnumFromJSONTyped(json, false);
}

export function StageEnumFromJSONTyped(json: any, ignoreDiscriminator: boolean): StageEnum {
    return json as StageEnum;
}

export function StageEnumToJSON(value?: StageEnum | null): any {
    return value as any;
}
